module/zfs/metaslab.c

```diff
@@ -268,14 +300,31 @@ metaslab_class_validate(metaslab_class_t *mc)
 	return (0);
 }
 
-void
+static void
 metaslab_class_space_update(metaslab_class_t *mc, int64_t alloc_delta,
     int64_t defer_delta, int64_t space_delta, int64_t dspace_delta)
 {
 	atomic_add_64(&mc->mc_alloc, alloc_delta);
 	atomic_add_64(&mc->mc_deferred, defer_delta);
 	atomic_add_64(&mc->mc_space, space_delta);
 	atomic_add_64(&mc->mc_dspace, dspace_delta);
+
+	spa_t *spa = mc->mc_spa;
+	uint64_t allocated = mc->mc_alloc;
+
+	if (mc == spa_normal_class(spa)) {
+		MS_CLASS_STAT(normal_allocated) = allocated;
+		if (allocated > MS_CLASS_STAT(normal_highest_allocated))
+			MS_CLASS_STAT(normal_highest_allocated) = allocated;
+	} else if (mc == spa_special_class(spa)) {
+		MS_CLASS_STAT(special_allocated) = allocated;
+		if (allocated > MS_CLASS_STAT(special_highest_allocated))
+			MS_CLASS_STAT(special_highest_allocated) = allocated;
+	} else if (mc == spa_log_class(spa)) {
+		MS_CLASS_STAT(slog_allocated) = allocated;
+		if (allocated > MS_CLASS_STAT(slog_highest_allocated))
+			MS_CLASS_STAT(slog_highest_allocated) = allocated;
+	}
```

Comment:
Yep. I switched it to a atomic_cas_64 to avoid undoing a higher value from another thread.